pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2164,6 +2168,190 @@ public boolean instanceExists(String instanceName) {
+      throw new RuntimeException(errorMsg, e);
+    }
+
+    // Only successful attempt can reach here
+    LOGGER.info("startBatchUpload is successfully processed. (tableNameWithType = {}, segmentsFrom = {}, "
+        + "segmentsTo = {}, batchId = {})", tableNameWithType, segmentsFrom, segmentsTo, batchId);
+    return batchId;
+  }
+
+  /**
+   * Computes the end batch upload phase
+   *
+   * 1. Compute validation
+   * 2. Update the lineage entry state to "COMPLETED" and write metadata to the property store
+   *
+   * Update is done with retry logic along with read-modify-write block for achieving atomic update of the lineage
+   * metadata.
+   *
+   * @param tableNameWithType
+   * @param batchId
+   */
+  public void endBatchUpload(String tableNameWithType, String batchId) {
+    // Check that the batch id is valid
+    if (batchId == null || batchId.isEmpty()) {
+      throw new IllegalArgumentException("'batchId' cannot be null or empty");
+    }
+
+    try {
+      DEFAULT_RETRY_POLICY.attempt(() -> {
+        // Fetch the segment lineage metadata
+        ZNRecord segmentLineageZNRecord =
+            SegmentLineageAccessHelper.getSegmentLineageZNRecord(_propertyStore, tableNameWithType);
+        SegmentLineage segmentLineage;
+        int expectedVersion = -1;
+        if (segmentLineageZNRecord == null) {
+          String errorMsg = String
+              .format("Segment lineage does not exist. (tableNameWithType = '%s', batchId = '%s')", tableNameWithType,
+                  batchId);
+          throw new IllegalArgumentException(errorMsg);
+        } else {
+          segmentLineage = SegmentLineage.fromZNRecord(segmentLineageZNRecord);
+          expectedVersion = segmentLineageZNRecord.getVersion();
+        }
+
+        // Look up the lineage entry based on the batch id
+        LineageEntry lineageEntry = segmentLineage.getLineageEntry(batchId);
+        if (lineageEntry == null) {
+          String errorMsg =
+              String.format("Invalid batch id (tableName='%s', batchId='%s')", tableNameWithType, batchId);
+          throw new IllegalArgumentException(errorMsg);
+        }
+
+        // Check that all the segments from 'segmentsTo' exist in the table
+        Set<String> segmentsForTable = new HashSet<>(getSegmentsFor(tableNameWithType));
+        if (!segmentsForTable.containsAll(lineageEntry.getSegmentsTo())) {
+          String errorMsg = String.format(
+              "Not all segments from 'segmentsTo' are available in the table. (tableName = '%s', segmentsTo = '%s', "
+                  + "segmentsFromTable = '%s')", tableNameWithType, lineageEntry.getSegmentsTo(), segmentsForTable);
+          throw new IllegalArgumentException(errorMsg);
+        }
+
+        // NO-OPS if the entry is already completed
+        if (lineageEntry.getState() == LineageEntryState.COMPLETED) {
+          LOGGER.info("Lineage entry state is already COMPLETED. Nothing to update. (tableNameWithType={}, batchId={})",
```

Comment:
Probably WARN? This is unexpected. Also, move this in front of the segments check